### PR TITLE
Add Github Action to set version-info's announcement_url

### LINKED ISSUE #33634: Write Github Action to add new URL

### PR DESCRIPTION
**Pull request by tsmacdonald**
[Fixes #33634]

- adds a `announcement_url` property to the latest release in the version-info.json via a manually invoked github action

Screen Shot 2023-09-20 at 11 00 32 AM

Screen Shot 2023-09-20 at 11 00 51 AM

requires three new (non-secret) github repository variables:
```
AWS_REGION
AWS_S3_STATIC_BUCKET
AWS_CLOUDFRONT_STATIC_ID
```

### CODE REVIEW

**Comment by iethree** [comment]
this feels like a hack.

1) why not make this part of the [release script](https://github.com/metabase/metabase/blob/c1d88baef0de186026b946b74ea45a461564e594/bin/build/src/release/version_info.clj) where we write version-info.json in the first place?
2) there's a [separate](https://static.metabase.com/version-info-ee.json) `version-info-ee.json` we need to handle too
3) if this is meant to be triggered manually, I think you need a `workflow_dispatch` [trigger](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#workflow_dispatch).

**Comment by npretto** [comment]
@iethree I can answer #1, the blog post is posted after the actual release so we can't publish the announcement_url in version-info at that time (it's going to trigger the What's new notification with a link to it for people who do upgrade)

**Comment by iethree** [comment]
> @iethree I can answer 1, the blog post is posted after the actual release so we can't publish the announcement_url in version-info at that time (it's going to trigger the What's new notification with a link to it for people who do upgrade)

ok, that makes sense, but it would be nice to have all the code that updates those files co-located rather than having a totally separate action mutating it, which may not be at all clear to someone editing the release script.

**Comment by tsmacdonald** [comment]
>ok, that makes sense, but it would be nice to have all the code that updates those files co-located rather than having a totally separate action mutating it, which may not be at all clear to someone editing the release script.
 
 yeah, I draft-i-micated this since I'm redoing it as some extra Clojure in the existing `bin/build/.../version_info.clj` file. It doesn't spark joy to download the entire repository for what's going to be less than a kilobyte of code, but Roman says it's worth it.

**Comment by iethree** [comment]
> > ok, that makes sense, but it would be nice to have all the code that updates those files co-located rather than having a totally separate action mutating it, which may not be at all clear to someone editing the release script.
> 
> yeah, I draft-i-micated this since I'm redoing it as some extra Clojure in the existing `bin/build/.../version_info.clj` file. It doesn't spark joy to download the entire repository for what's going to be less than a kilobyte of code, but Roman says it's worth it.

you can modify the checkout action to only checkout certain files.
https://github.com/actions/checkout#fetch-only-a-single-file

File: bin/build/src/release/version_info.clj
Comment (tsmacdonald) [inline bin/build/src/release/version_info.clj:73]: I hate boolean flags like this but couldn't come up with a better idea. It somehow seemed grosser to parameterize this function with the copy function (i.e., `(defn upload-version-info! [edition copy-fn] ...)`)

File: bin/build/src/release/version_info.clj
Comment (iethree) [inline bin/build/src/release/version_info.clj:18]: maybe not the most elegant testing suggestion:

```suggestion
  (case edition
    :oss "version-info-test.json"
    :ee  "version-info-ee-test.json"
```

Comment (iethree) [reply bin/build/src/release/version_info.clj:18]: oh yeah, won't actually work because `workflow_dispatch` actions only work on the master branch 😢

**Comment by qnkhuat** [comment]
I'm new to the release process so would love to see some clarification on what's going on here.
- What is the purpose of this PR?
- How is the command gonna be used?
- Some explanation of these artifacts would be a great help.

**Comment by NevRA** [comment]
@qnkhuat 

> * What is the purpose of this PR?

We want to have a way to add the link to the latest blog post for the new version upgrade banner https://www.notion.so/Provide-a-heads-up-when-Metabase-is-upgraded-with-what-s-new-44a28b55624f48e28dc3f616ce7cda0f
> * How is the command gonna be used?

Run manually (had thread about it with products and grow team) via Github Actions when blog published

> * Some explanation of these artifacts would be a great help.

This is our list of released versions we update during the release process.

**Comment by tsmacdonald** [comment]
@qnkhuat to give some more tech detail, the FE is going to show a "Welcome to Version 48, click here to read about the awesome new features" message, but we need to know what URL to point that too. We want Growth to own providing the URL, and it won't necessarily exist when we build the gold version, so the build manager can't just include it then.

So the plan is to have a manual Github Action that Growth uses to supply the URL. We pull down [the current version-info blob](http://static.metabase.com/version-info.json), add the `latest.announcement_url` value, and then push it back up.

There was a lot of existing Clojure code that added a new version to `version-info` as part of the build process, so that's the code that I recycled/extended.

File: bin/build/src/metabuild_common/aws.clj
Comment (qnkhuat) [inline bin/build/src/metabuild_common/aws.clj:23]: could you explain why we need this version? why can't we use the existing on with profile?

File: bin/build/src/release/version_info.clj
Comment (qnkhuat) [inline bin/build/src/release/version_info.clj:20]: where do we use these `-test` artifacts?

**Comment by iethree** [comment]
I have a metabase fork that's fully set up for CI now so I can test this. will report back

**Comment by iethree** [comment]
OK, to make up for being the world's most negligent code reviewer, I fixed this up myself and pushed it.

- I went back to the version that was purely github-actions based, instead of putting more in the clojure release script [that we're deprecating](https://github.com/metabase/metabase/pull/33938)
- I also added some validation
- the main advantage of this is that it neither blocks nor is blocked by https://github.com/metabase/metabase/pull/33938

I also tested it in my fork here:
oss: [action run](https://github.com/iethree/metabase/actions/runs/6250369133)  | [published artifact](http://iethree-static.metabase.com.s3.us-east-1.amazonaws.com/version-info.json)
ee: [action run](https://github.com/iethree/metabase/actions/runs/6250367999)  | [published artifact](http://iethree-static.metabase.com.s3.us-east-1.amazonaws.com/version-info-ee.json)

File: .github/workflows/set-announce-url.yml
Comment (nemanjaglumac) [inline .github/workflows/set-announce-url.yml:28]:
```suggestion
      - name: Configure AWS credentials
        uses: aws-actions/configure-aws-credentials@v3
```

They've just released v4 recently, but let's stick with the `v3` for a while.

File: .github/workflows/set-announce-url.yml
Comment (nemanjaglumac) [inline .github/workflows/set-announce-url.yml:45]: Why do we use `fetch` here when we can do it using the provided `aws` cli tool, like in the "Upload" step?
This also has a hard-coded region which might change.

File: .github/workflows/set-announce-url.yml
Comment (nemanjaglumac) [inline .github/workflows/set-announce-url.yml:23]:
```suggestion
            if (!announcementUrl.startsWith('http')) {
```

File: .github/workflows/set-announce-url.yml
Comment (nemanjaglumac) [inline .github/workflows/set-announce-url.yml:56]:
```suggestion
          aws s3 cp version-info.json s3://${{ vars.AWS_S3_STATIC_METABASE_BUCKET }}/version-info.json
          aws s3 cp version-info-ee.json s3://${{ vars.AWS_S3_STATIC_METABASE_BUCKET }}/version-info-ee.json
```

or something along these lines. We need to be able to distinguish between `static` and `downloads` in vars.

**Review by nemanjaglumac — CHANGES_REQUESTED** [review]
Left some inline comments.
Overall looks good, but if would be great if we had some tests, and/or validations in case something goes wrong.

For example, what's the guarantee that the downloaded json is not garbled? We risk uploading and overwriting a perfectly valid JSON with a malformed one.

File: .github/workflows/set-announce-url.yml
Comment (iethree) [reply .github/workflows/set-announce-url.yml:45]: 1. I updated the hard-coded region to use a variable. 👍 
2. I think it makes sense to use fetch because it ensures we're getting a publicly accessible object, and I think it's simpler because we can fetch and parse all at once instead of download to disk, load into js from the filesystem, then parse.

**Review by iethree — APPROVED** [review]

File: .github/workflows/set-announce-url.yml
Comment (nemanjaglumac) [inline .github/workflows/set-announce-url.yml:28]: Nit:
```suggestion
          aws-region: ${{ vars.AWS_REGION }}
```

**Review by nemanjaglumac — APPROVED** [review]
Looks good.
Thanks for addressing review comments.

**Comment by tsmacdonald** [comment]
@iethree I went to [your `version-info.json` fork](http://iethree-static.metabase.com.s3.us-east-1.amazonaws.com/version-info.json) and didn't see the new announcement-url in there. Did it just get overwritten or something?

Assuming that's okay, LGTM, feel free to hit the merge button when you like

**Comment by iethree** [comment]
> @iethree I went to [your `version-info.json` fork](http://iethree-static.metabase.com.s3.us-east-1.amazonaws.com/version-info.json) and didn't see the new announcement-url in there. Did it just get overwritten or something?
> 
> Assuming that's okay, LGTM, feel free to hit the merge button when you like

yeah i overwrote it. will run one more test there before merging

**Pull request merged**
